ratelimits/limiter.go

```diff
@@ -238,90 +164,212 @@ func (l *Limiter) Spend(ctx context.Context, txn Transaction) (*Decision, error)
 // instance, if a bucket has a maximum capacity of 10 and currently has 5
 // requests remaining, a refund request of 7 will result in the bucket reaching
 // its maximum capacity of 10, not 12.
 func (l *Limiter) Refund(ctx context.Context, txn Transaction) (*Decision, error) {
-	if txn.cost <= 0 {
-		return nil, ErrInvalidCost
-	}
-
-	limit, err := l.getLimit(txn.limit, txn.bucketKey)
-	if err != nil {
-		if errors.Is(err, errLimitDisabled) {
-			return disabledLimitDecision, nil
-		}
-		return nil, err
+	if txn.allowOnly() {
+		return allowedDecision, nil
 	}
-
 	// Remove cancellation from the request context so that transactions are not
 	// interrupted by a client disconnect.
 	ctx = context.WithoutCancel(ctx)
 	tat, err := l.source.Get(ctx, txn.bucketKey)
 	if err != nil {
 		return nil, err
 	}
-	d := maybeRefund(l.clk, limit, tat, txn.cost)
-	if !d.Allowed {
-		// The bucket is already at maximum capacity.
-		return d, nil
+	d := maybeRefund(l.clk, txn.limit, tat, txn.cost)
+	if tat == d.newTAT || txn.checkOnly() {
+		return maybeRefund(l.clk, txn.limit, tat, 0), nil
 	}
-	return d, l.source.Set(ctx, txn.bucketKey, d.newTAT)
-}
-
-// Reset resets the specified bucket to its maximum capacity. The new bucket
-// state is persisted to the underlying datastore before returning.
-func (l *Limiter) Reset(ctx context.Context, bucketId BucketId) error {
-	// Remove cancellation from the request context so that transactions are not
-	// interrupted by a client disconnect.
-	ctx = context.WithoutCancel(ctx)
-	return l.source.Delete(ctx, bucketId.bucketKey)
+	if d.Allowed {
+		// Persist the new bucket state.
+		return d, l.source.Set(ctx, txn.bucketKey, d.newTAT)
+	}
+	// Bucket is already full.
+	return d, nil
 }
 
-// initialize creates a new bucket and sets its TAT to now, which is equivalent
-// to a full bucket. The new bucket state is persisted to the underlying
-// datastore before returning.
-func (l *Limiter) initialize(ctx context.Context, rl limit, txn Transaction) (*Decision, error) {
-	d := maybeSpend(l.clk, rl, l.clk.Now(), txn.cost)
+// BatchRefund attempts to refund all or some of the costs to the provided
+// buckets' capacities. Non-existent buckets will NOT be initialized. The new
+// bucket state is persisted to the underlying datastore, if applicable, before
+// returning. Spend-only Transactions are assumed to be refundable. Check-only
+// Transactions are never refunded. The following rules are applied to merge the
+// Decisions for each Transaction into a single batch Decision:
+//   - Allowed is true if all Transactions where check is true were allowed,
+//   - RetryIn and ResetIn are the largest values of each across all Decisions,
+//   - Remaining is the smallest value of each across all Decisions, and
+//   - Decisions resulting from spend-only Transactions are never merged.
+func (l *Limiter) BatchRefund(ctx context.Context, txns []Transaction) (*Decision, error) {
```

Comment:
It looks like there's some duplicated code between `Refund` and `BatchRefund`. That could be improved by having `Refund` call `BatchRefund` with a single-element slice.